src/rbindlist.c

```diff
@@ -514,20 +514,24 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
         if (w==-1 || !length(thisCol=VECTOR_ELT(li, w))) {  // !length for zeroCol warning above; #1871
           writeNA(target, ansloc, thisnrow);  // writeNA is integer64 aware and writes INT64_MIN
         } else {
-          if ((TYPEOF(target)==VECSXP || TYPEOF(target)==EXPRSXP) && TYPEOF(thisCol)!=TYPEOF(target)) {
-            // do an as.list() on the atomic column; #3528
-            thisCol = PROTECT(coerceVector(thisCol, TYPEOF(target))); nprotect++;
+          if (maxType==VECSXP) {
+            thisCol = PROTECT(coerceAsList(thisCol, thisnrow)); nprotect++;
           }
+
           // else coerces if needed within memrecycle; with a no-alloc direct coerce from 1.12.4 (PR #3909)
           const char *ret = memrecycle(target, R_NilValue, ansloc, thisnrow, thisCol, 0, -1, idcol+j+1, foundName);
           if (ret) warning(_("Column %d of item %d: %s"), w+1, i+1, ret);
           // e.g. when precision is lost like assigning 3.4 to integer64; test 2007.2
           // TODO: but maxType should handle that and this should never warn
+       
+          // memrecycle incorrectly adds integer64 class to whole list if any element has class integer64
+          if (maxType == VECSXP && INHERITS(target, char_integer64)) 
+            setAttrib(target, R_ClassSymbol, R_NilValue);
```

Comment:
Ideally would patch in memrecycle

Reply:
I have so far not been able to track down where or why in memrecycle this happens (both target and thisCol do not inherit char_integer64), and this class assignment to the full list does not seem to happen if I add use other classes (e.g. assigning class "foo" to a character vector column, then rbinding to a list column).

Reply:
early copyMostAttrib statement was the culprit. now skipped when maxType=VECSXP